### PR TITLE
test/bugfix: persistent receive with anysource.

### PR DESCRIPTION
**Pull request by danghvu**
This test hangs ch4/ofi/psm2 on an omnipath, found while testing graph500 new reference implementation.
This PR is for checking with other ch3/ch4 and fixing.
Update: Two independent bugs are found:
1) The ch4 MPID_Startall originally did startall for both SHM and NM layer if `MPI_ANY_SOURCE` is specified, which then posted `Irecv` twice for SHM, one in the `MPIDI_SHM_mpi_startall`, and one in ch4 `MPID_Irecv` of `MPIDI_NM_mpi_startall`.
2) When AM layer of OFI is used, Irecv can complete immediately if the message is already arrived (in unexpected queue), in this case the implementation did not clear the partner request causing hang (because `MPID_Waitall` only check the partner request).

### CODE REVIEW

**Comment by danghvu** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch4/most

**Comment by danghvu** [comment]
test:jenkins/ch4/most

**Comment by danghvu** [comment]
test:jenkins/ch4/most

**Comment by danghvu** [comment]
test:jenkins/ch4/most

**Comment by danghvu** [comment]
test:jenkins/ch4/most

**Comment by danghvu** [comment]
test:jenkins/ch4/ofi

**Comment by danghvu** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch4/most

**Comment by danghvu** [comment]
test:jenkins/ch4/most

**Comment by danghvu** [comment]
test:jenkins/ch4/most

**Comment by danghvu** [comment]
@raffenet please review this when you have a chance, I left the fixup to show previous jenkins test results. There seems to be two different  bugs, one will [hang ch4/ofi](https://jenkins.mpich.org/job/mpich-review-ch4-ofi/1434/) , and one will [hang ch4/ofi/AM](https://jenkins.mpich.org/job/mpich-review-ch4-ofi/1435/) after fixing the first one.

File: test/mpi/pt2pt/testlist.def [hunk 1]
```diff
@@ -47,3 +47,9 @@ huge_dupcomm 2
 dtype_send 2
 recv_any 2
 irecv_any 2
+precv_anysrc 2 arg=-msg_size=64 timeLimit=60
```
Comment (raffenet) [inline test/mpi/pt2pt/testlist.def]: Seems like a little overkill to split message size out like this. Can you just iterate over these inside the test case? You can add debug output to show when each iteration is complete so it's easy to investigate when there is a failure.

**Comment by danghvu** [comment]
test:jenkins/ch4/most
test:jenkins/ch3/tcp

**Comment by danghvu** [comment]
@raffenet I saw some PR related to any_source (e.g. #3199) , but not persistent, should we push this so that the other PR can pick up new test?

**Comment by hzhou** [comment]
Can this PR be updated?

File: src/mpid/ch4/src/ch4_recv.h [hunk 2]
```diff
@@ -129,26 +129,13 @@ MPL_STATIC_INLINE_PREFIX int MPID_Recv_init(void *buf,
...
-        }
-
-        mpi_errno =
-            MPIDI_NM_mpi_recv_init(buf, count, datatype, rank, tag, comm, context_offset, av,
-                                   &(MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(*request)));
-
-        if (mpi_errno != MPI_SUCCESS) {
-            MPIR_ERR_POP(mpi_errno);
+        /* NM_Irecv for ANY_SOURCE will also post a receive for SHM. */
+        mpi_errno = MPIDI_NM_mpi_recv_init(buf, count, datatype, rank, tag,
```
Comment (hzhou) [inline src/mpid/ch4/src/ch4_recv.h]: What if SHM and NM need initialize differently?

I would rather see a fix that prevents `MPI_Start` double post than a solution that restricts on functionality.

File: src/mpid/ch4/src/ch4_startall.h [hunk 3]
```diff
@@ -28,18 +28,7 @@ MPL_STATIC_INLINE_PREFIX int MPID_Startall(int count, MPIR_Request * requests[])
...
-            if (mpi_errno == MPI_SUCCESS) {
-                mpi_errno = MPIDI_NM_mpi_startall(1, &MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(req));
-                MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(req->u.persist.real_request) =
-                    MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(req)->u.persist.real_request;
-                MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER
-                                                     (req)->u.persist.real_request) =
-                    req->u.persist.real_request;
-            }
-        } else if (MPIDI_CH4I_REQUEST(req, is_local))
+        if (MPIDI_CH4I_REQUEST(req, is_local))
```
Comment (hzhou) [inline src/mpid/ch4/src/ch4_startall.h]: Is the change here meant to prevent double launching? Then we don't need the fix in `ch4_recv.h`, right?

**Comment by raffenet** [comment]
Hmm, #3914 claimed to fix persistent anysrc receive in ch4. Perhaps this is no longer needed. I don't think the original author is going to update it, in any case.

**Comment by hzhou** [comment]
It might be a good idea to import the test case at least, then we can verify if the problem still exist in current master.

**Comment by pavanbalaji** [comment]
The tests have been moved to https://github.com/pmodels/mpich/pull/4304, so this PR can be closed.

**Pull request closed without merging**